sprokit/src/sprokit/python/util/python_gil.h

```diff
@@ -36,37 +36,126 @@
 #include <vital/noncopyable.h>
 
 #include <sprokit/python/util/python.h>
+#include <sprokit/python/util/python_exceptions.h>
+
+#include <pybind11/pybind11.h>
+#include <pybind11/cast.h>
+
+#include <memory>
 
 /**
  * \file python_gil.h
  *
- * \brief RAII class for grabbing the Python GIL.
+ * \brief Helper utilities for grabbing the Python GIL.
  */
 
 namespace sprokit {
 namespace python {
 
 /**
- * \class python_gil python_gil.h <sprokit/python/util/python_gil.h>
+ * \class python_gil_cond_release python_gil.h <sprokit/python/util/python_gil.h>
  *
- * \brief Grabs the Python GIL and uses RAII to ensure it is released.
+ * \brief Releases the python gil if held by this thread for the duration this
+ * class is in scope, but only if this is a valid python thread otherwise this
+ * is a no-op.
  */
-class SPROKIT_PYTHON_UTIL_EXPORT python_gil
+class SPROKIT_PYTHON_UTIL_EXPORT python_gil_cond_release
   : private kwiver::vital::noncopyable
 {
   public:
     /**
      * \brief Constructor.
      */
-    python_gil();
+    python_gil_cond_release();
     /**
      * \brief Destructor.
      */
-    ~python_gil();
+    ~python_gil_cond_release();
+
+  private:
+    std::unique_ptr< pybind11::gil_scoped_release > rel;
+};
+
+/**
+ * \class Record settings relating to the GIL
+ */
+class SPROKIT_PYTHON_UTIL_EXPORT python_gil_settings
+{
+  public:
+    static bool cycle_gil_lock();
+    static void set_cycle_option( bool opt = true );
+
   private:
-    PyGILState_STATE const state;
+    python_gil_settings() : cycle_gil_flag( false ) {}
+    ~python_gil_settings() {}
+
+    static python_gil_settings instance;
+    bool cycle_gil_flag;
+
+  public:
+    python_gil_settings( python_gil_settings const& ) = delete;
+    void operator=( python_gil_settings const& )      = delete;
 };
 
+/**
+ * \brief Helper macros for warningless pybind11 GIL scoped acquisitions
+ */
+#define SPROKIT_SCOPED_GIL_ACQUIRE_START                                          \
+  {                                                                               \
+    pybind11::gil_scoped_acquire acquire;                                         \
+                                                                                  \
+    (void) acquire;
+
+#define SPROKIT_SCOPED_GIL_ACQUIRE_END                                            \
+  }
+
```

Comment:
Having a start and an end scope macro hides the brackets from the user and the editor so it is not clear what is really going on. Somebody could assume that the scope is not really a bracketed scope but some other mechanism and put the SPROKIT_SCOPED_GIL_ACQUIRE_END call at a different block level.

It would be more obvious if the code from these macros were just in the calling code. A macro could be used to hide the pybind11 call if desired e.g. SPROKIT_SCOPED_GIL_ACQUIRE();  or something to make it look like a call.